src/pytest_mock_resources/action.py

```diff
@@ -0,0 +1,32 @@
+from __future__ import annotations
+
+import abc
+from typing import ClassVar, Iterable, Optional
+
+
+class AbstractAction(metaclass=abc.ABCMeta):
+    fixtures: ClassVar[tuple[str, ...]] = ()
+    static_safe: ClassVar[bool] = False
+
+    @abc.abstractmethod
+    def apply(self, conn):
+        """Execute an action against the provided fixture connection."""
+
+
+def validate_actions(actions, *, fixture: Optional[str], additional_types: Iterable = ()):
```

Comment:
with that being said, it's mildly unfortunate that it forces fixtures themselves to push down their fixture "name" (i.e. "mysql") in through the functions that might validate actions.

Right now that means that i.e. redshift/mysql/etc are sending a `fixture="redshift"` string in through the postgres machinery that they're based on, which let them produce the correct error messages when misused.

idk. it seems worth to me, but 🤷

Reply:
Does this ever get surfaced to the user? If not, its not a terrible deal. If so, could we infer the type by the fn name/passed in args?

Reply:
it does not. We "need" to propagate it downward from inside fixtures (using these utilities) so that the error message references the correct fixture and validates the correct available values.